meteor/client/ui/Shelf/PieceCountdownPanel.tsx

```diff
@@ -0,0 +1,120 @@
+import { RundownTiming, TimingEvent } from '../RundownView/RundownTiming/RundownTiming'
+import { RundownPlaylist } from '../../../lib/collections/RundownPlaylists'
+import { PieceInstance } from '../../../lib/collections/PieceInstances'
+import { VTContent } from '@sofie-automation/blueprints-integration'
+interface IPieceCountdownPanelProps {
+	visible?: boolean
+	layout: RundownLayoutBase
+	panel: RundownLayoutPieceCountdown
+	playlist: RundownPlaylist
+}
+
+interface IPieceCountdownPanelTrackedProps {
+	livePieceInstance?: PieceInstance
+}
+
+interface IState {
+	displayTimecode: number
+}
+
+export class PieceCountdownPanelInner extends MeteorReactComponent<
+	IPieceCountdownPanelProps & IPieceCountdownPanelTrackedProps,
+	IState
+> {
+	constructor(props) {
+		super(props)
+		this.state = {
+			displayTimecode: 0,
+		}
+		this.updateTimecode = this.updateTimecode.bind(this)
+	}
+
+	componentDidMount() {
+		window.addEventListener(RundownTiming.Events.timeupdate, this.updateTimecode)
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener(RundownTiming.Events.timeupdate, this.updateTimecode)
+	}
+
+	updateTimecode(e: TimingEvent) {
+		let timecode = 0
+		if (this.props.livePieceInstance && this.props.livePieceInstance.startedPlayback) {
+			const vtContent = this.props.livePieceInstance.piece.content as VTContent | undefined
+			const sourceDuration = vtContent?.sourceDuration || 0
+			const startedPlayback = this.props.livePieceInstance.startedPlayback
+			if (startedPlayback && sourceDuration > 0) {
+				timecode = e.detail.currentTime - (startedPlayback + sourceDuration)
+			}
+		}
+		if (this.state.displayTimecode != timecode) {
+			this.setState({
+				displayTimecode: timecode,
+			})
+		}
+	}
+
+	render() {
+		return (
+			<div
+				className="piece-countdown-panel"
+				style={_.extend(
+					RundownLayoutsAPI.isDashboardLayout(this.props.layout)
+						? dashboardElementPosition(this.props.panel as DashboardLayoutPartCountdown)
+						: {},
+					{
+						visibility: this.props.visible ? 'visible' : 'hidden',
+					}
+				)}>
+				<span
+					className={ClassNames('piece-countdown-panel__timecode', {
+						overtime: !!(Math.floor(this.state.displayTimecode / 1000) > 0),
+					})}>
+					{RundownUtils.formatDiffToTimecode(
+						this.state.displayTimecode || 0,
+						true,
+						false,
+						true,
+						false,
+						true,
+						'',
+						false,
+						true
+					)}
+				</span>
+			</div>
+		)
+	}
+}
+
+export const PieceCountdownPanel = withTracker<IPieceCountdownPanelProps, IState, IPieceCountdownPanelTrackedProps>(
+	(props: IPieceCountdownPanelProps & IPieceCountdownPanelTrackedProps) => {
+		const unfinishedPieces = getUnfinishedPieceInstancesReactive(props.playlist.currentPartInstanceId)
+		const livePieceInstance: PieceInstance | undefined =
+			props.panel.sourceLayerIds && props.panel.sourceLayerIds.length
+				? _.find(_.flatten(_.values(unfinishedPieces)), (piece: PieceInstance) => {
```

Comment:
I think we are trying to limit our reliance on underscore, this could easily be replaced with `Object.values().flat().find()` with better performance at type security.

Reply:
`Array.prototype.flat()` is an es2019 feature and looks like we're targeting es2018

Reply:
Oh, it is & we are? I didn't realize that. Well, the question then is if it makes sense to just upgrade to es2019 (I think with Node 12 we can), but I guess the easiest option is to use `_.flatten`, but replace `_.values` and `_.find` with native alternatives for this PR.

Reply:
https://node.green/#ES2019 yeah we should be good to target es2019
Im pretty sure meteor is doing babel transpiling anyway, so maybe we could even switch that to target esnext

Reply:
I replaced `values` and `find`. Shall moving to es2019 be made in another PR, perhaps with refactoring any other occurrences of `_.flatten` etc.?